### ISSUE TITLE
Checkpoint.restore doesn't restore Dataset iterator state when Dataset contains shuffle()

### ISSUE THREAD

**Issue by kaisuke** [tensorflow/tensorflow#34020]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.0.0-rc2-26-g64c3d38 2.0.0
- Python version: 3.5.2
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: not installed
- GPU model and memory: no GPU

**Describe the current behavior**
The code at the end results in the following output.
```
2019-11-05 13:54:16.140994: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-11-05 13:54:16.149389: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2397225000 Hz
2019-11-05 13:54:16.151608: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x50ebc70 executing computations on platform Host. Devices:
2019-11-05 13:54:16.151636: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): Host, Default Version
tf.Tensor(72, shape=(), dtype=int64)
tf.Tensor(83, shape=(), dtype=int64)
tf.Tensor(19, shape=(), dtype=int64)
Saved to /tmp/x/ckpt-1
tf.Tensor(74, shape=(), dtype=int64)
tf.Tensor(33, shape=(), dtype=int64)
tf.Tensor(93, shape=(), dtype=int64)
Restored from /tmp/x/ckpt-1
tf.Tensor(21, shape=(), dtype=int64)
tf.Tensor(0, shape=(), dtype=int64)
tf.Tensor(8, shape=(), dtype=int64)
```

**Describe the expected behavior**
After restoring from the checkpoint, I expect the iterator to return the same elements as it did after saving the checkpoint.  I.e.
```
2019-11-05 13:54:16.140994: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-11-05 13:54:16.149389: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2397225000 Hz
2019-11-05 13:54:16.151608: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x50ebc70 executing computations on platform Host. Devices:
2019-11-05 13:54:16.151636: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): Host, Default Version
tf.Tensor(72, shape=(), dtype=int64)
tf.Tensor(83, shape=(), dtype=int64)
tf.Tensor(19, shape=(), dtype=int64)
Saved to /tmp/x/ckpt-1
tf.Tensor(74, shape=(), dtype=int64)
tf.Tensor(33, shape=(), dtype=int64)
tf.Tensor(93, shape=(), dtype=int64)
Restored from /tmp/x/ckpt-1
tf.Tensor(74, shape=(), dtype=int64)
tf.Tensor(33, shape=(), dtype=int64)
tf.Tensor(93, shape=(), dtype=int64)
```

**Code to reproduce the issue**
```py
import tensorflow as tf
ds = tf.data.Dataset.range(100).shuffle(100, seed=42, reshuffle_each_iteration=False)
it = iter(ds)
ckpt = tf.train.Checkpoint(foo=it)
mgr = tf.train.CheckpointManager(ckpt, '/tmp/x', max_to_keep=3)
for _ in range(3): print(next(it))
mgr.save()
print("Saved to {}".format(mgr.latest_checkpoint))
for _ in range(3): print(next(it))
ckpt.restore(mgr.latest_checkpoint)
print("Restored from {}".format(mgr.latest_checkpoint))
for _ in range(3): print(next(it))
```
FWIW, I get the expected result if I remove `.shuffle(...)`.

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by oanush** [comment]
Issue replicating for TF-2.0, kindly find the [gist](https://colab.sandbox.google.com/gist/oanush/571af19a2605e0b63938207f55803361/untitled26.ipynb) of colab.Thanks!

**Comment by gowthamkpr** [comment]
@kaisuke Checkpoint.save and Checkpoint.restore write and read object-based checkpoints, in contrast to TensorFlow 1.x's tf.compat.v1.train.Saver which writes and reads variable.name based checkpoints. Object-based checkpointing saves a graph of dependencies between Python objects (Layers, Optimizers, Variables, etc.) with named edges, and this graph is used to match variables when restoring a checkpoint. It can be more robust to changes in the Python program, and helps to support restore-on-create for variables.

This is the reason why you are observing changes when you are using .shuffle() method. For more information, you can go through the following [link](https://www.tensorflow.org/api_docs/python/tf/train/Checkpoint). Thanks!

**Comment by jsimsa** [comment]
@aaudiber could you please take a look? thanks

**Comment by aaudiber** [comment]
I was able to reproduce the issue. It appears that some state is not properly reset when restoring an active shuffle iterator from a checkpoint. I have a fix in-flight.
